extension/packagemanagerextension/nbextension/methods.js

```diff
@@ -0,0 +1,146 @@
+
+define([
+    'jquery',
+    './common',
+    './api',
+], function ($, common, api) {
+    "use strict";
+
+
+    /*
+    This function returns the latest version of a package.
+    */
+
+    function checknewversion(updates, packagename, pkgversion) {
+        for (let element of updates) {
+            if (element.name === packagename) {
+                return element.version;
+            }
+        }
+        return pkgversion;
+    }
+
+    /*
+    This function lists down all the updates and requires a user confirmation.
+    */
+
+    async function updatepkg() {
+        $('#updatebtn').unbind();
+        $('#updatebtn').click(async function () {
+            let selectedPackages = common.get_selected_packages();
+            let url = location.href.split('/');
+            let project = 'SWAN_projects/' + url[url.length - 1];
+            let html = "<p> The following packages are about to be updated: </p> </br>";
+            $('#updatebtn').toggleClass('fa-wrench fa-spinner').addClass('fa-spin');
+            let resp = await api.checkupdate(project);
+            $('#updatebtn').toggleClass('fa-wrench fa-spinner').removeClass('fa-spin');
+            let updates = resp.updates;
+            let packages = [];
+            let ct = 0;
+
+            if (selectedPackages.length > 0) {
+                html += "<ul>";
+                selectedPackages.forEach(element => {
+                    element = element.split('=');
+                    let pkg = element[0];
+                    let ver = element[1];
+                    let nver = checknewversion(updates, pkg, ver);
+                    if (nver != ver) {
+                        packages.push(pkg);
+                        html += "<li>";
+                        html += pkg + "    " + ver + " -> " + nver;
+                        html += "</li>";
+                        ct += 1;
+                    }
+                });
+                html += "</ul>";
+            }
+            else {
+                let list = $('.installed-values');
+                html += "<ul>";
+                for (let item of list) {
+                    let pkg = $(item).find('.value-name')[0].innerText;
```

Comment:
I suggested above to keep internal arrays with representations of these lists. But, even if you keep going through the DOM to get this info (since I said it was ok), you should at least create some function to abstract behaviour (you pass the css class name, and you get a list of pkg+version objects).

Reply:
And the code to show the update is exactly the same.. So if it's the same, you should create a function and call it whenever you need.
No code replication, please.

Reply:
You could even pass the list to the `common.confirm`. If list is empty, you show the message below. If not, you create an html representation. You would avoid repetition I mentioned in the above comment, plus move all html logic to one single place.

Reply:
And pick one... `..forEach(..)` or `for(let .. of ..)`.